File: backend/src/Loupe.Application/Locations/LocationResult.cs
using System.Globalization;
using Loupe.Domain.Locations;

namespace Loupe.Application.Locations;

public sealed record LocationResult(Guid Id, string Name, string? AddressLine1, string? AddressLine2, string? Locality, string? Region,
    string? PostalCode, string? Country, Coordinates? Coordinates, LocationSetting? Setting, string? ScoutingBrief, string? Notes,
    IReadOnlyList<LocationTagResult> Tags, IReadOnlyList<LocationImageResult> Images, Guid? CoverImageId, object? Report,
    string ReportStatus, DateTimeOffset CreatedAt, DateTimeOffset UpdatedAt, long Revision)
{
    public static LocationResult From(Location location) => new(location.Id, location.Name, location.AddressLine1, location.AddressLine2,
        location.Locality, location.Region, location.PostalCode, location.Country,
        location is { Latitude: { } latitude, Longitude: { } longitude } ? new Coordinates(Format(latitude), Format(longitude)) : null,
        location.Setting, location.ScoutingBrief, location.Notes,
        location.Tags.OrderBy(tag => tag.Name, StringComparer.OrdinalIgnoreCase).Select(tag => new LocationTagResult(tag.Name, tag.Category)).ToArray(),
        location.Images.OrderBy(image => image.Position).Select(image => LocationImageResult.From(location.Id, image)).ToArray(),
        location.CoverImageId, null, "None", location.CreatedAt, location.UpdatedAt, location.Revision);

    private static string Format(decimal value) => value.ToString("0.000000", CultureInfo.InvariantCulture);
}

File: frontend/projects/api/src/lib/location/location-result.ts
import { SavedScoutingReport } from '../scouting-report/scouting-report-result';

export type LocationSetting = 'Indoor' | 'Outdoor' | 'Mixed';
export type LocationReportStatus = 'None' | 'Queued' | 'Running' | 'Ready' | 'Outdated' | 'Failed';

export interface Coordinates {
  latitude: string;
  longitude: string;
}
export interface LocationTag {
  name: string;
  category: string | null;
}
export interface LocationImage {
  id: string;
  position: number;
  imageUrl: string;
  previewUrl: string;
  width: number;
  height: number;
}
export interface LocationResult {
  id: string;
  name: string;
  addressLine1: string | null;
  addressLine2: string | null;
  locality: string | null;
  region: string | null;
  postalCode: string | null;
  country: string | null;
  coordinates: Coordinates | null;
  setting: LocationSetting | null;
  scoutingBrief: string | null;
  notes: string | null;
  tags: LocationTag[];
  images: LocationImage[];
  coverImageId: string | null;
  report: SavedScoutingReport | null;
  reportStatus: LocationReportStatus;
  createdAt: string;
  updatedAt: string;
  revision: number;
}
export interface LocationSummary {
  id: string;
  name: string;
  locality: string | null;
  coverPreviewUrl: string | null;
  imageCount: number;
  reportStatus: LocationReportStatus;
  createdAt: string;
}
export interface LocationPage {
  items: LocationSummary[];
  nextCursor: string | null;
  totalCount: number;
}
